File: src/baekjoon/dp/b7579/b7579.java
package baekjoon.dp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class b7579 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] memories = new int[N];
        int[] costs = new int[N];
        int totalCost = 0;

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            memories[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            costs[i] = Integer.parseInt(st.nextToken());
            totalCost += costs[i];
        }

        // dp[c] = 비용 c를 들여 확보할 수 있는 최대 메모리
        int[] dp = new int[totalCost + 1];

        // 모든 앱을 순회
        for (int i = 0; i < N; i++) {
            int memory = memories[i];
            int cost = costs[i];

            // 역순으로 순회(핵심)
            for (int j = totalCost; j >= cost; j--) {
                // j 비용으로 i번째 앱을 비활성화하는 경우와 안하는 경우 중 최대 메모리 선택
                // dp[j] : i번째 앱을 비활성화 안 함
                // dp[j - cost] + memory : i번째 앱을 비활성화 함
                dp[j] = Math.max(dp[j], dp[j - cost] + memory);
            }
        }

        int minCost = 0;
        for (int i = 1; i <= totalCost; i++) {
            if (dp[i] >= M) {
                minCost = i;
                break;
            }
        }
        System.out.println(minCost);
    }
}

File: src/baekjoon/dp/b7579/b7579.py
import sys

# 빠른 입력을 위한 설정
input = sys.stdin.readline

def solve():
    N, M = map(int, input().split())
    memories = list(map(int, input().split()))
    costs = list(map(int, input().split()))

    total_cost = sum(costs)

    # dp[c] = 비용 c를 들여 확보할 수 있는 최대 메모리
    dp = [0] * (total_cost + 1)

    # 모든 앱을 순회
    for i in range(N):
        memory = memories[i]
        cost = costs[i]

        # 역순으로 순회 (핵심!)
        for j in range(total_cost, cost - 1, -1):
            # j 비용으로 i번째 앱을 비활성화하는 경우와 안하는 경우 중 최대 메모리 선택
            # dp[j] : i번째 앱을 비활성화 안 함
            # dp[j - cost] + memory : i번째 앱을 비활성화 함
            dp[j] = max(dp[j], dp[j - cost] + memory)

    # M 바이트 이상을 확보하는 최소 비용 찾기
    min_cost = 0
    for i in range(total_cost + 1):
        if dp[i] >= M:
            min_cost = i
            break

    print(min_cost)

solve()